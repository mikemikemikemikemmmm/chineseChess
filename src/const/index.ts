
import { TChessmanData, TFaction } from "../types"
export const width = window.innerWidth;
export const height = window.innerHeight;
export const gridWidth = height/14
export const chessBoardWidthGridNum = 8
export const chessBoardHeightGridNum = 9

/** 
 *  0: no chessman
 *  1: 車
 *  2: 馬
 *  3: 象
 *  4: 士
 *  5: 將
 *  6: 炮
 *  7: 兵
 * */
export const chessmanDataMap: { [props: string]: TChessmanData } = {
    '1': {   
        text: {
            red: '俥',
            black: '車'
        },
        moveRule: (boardData, startIndex, endIndex) => {
            if (startIndex.x === endIndex.x) {  // vertical move
                if (startIndex.y > endIndex.y) { // to up
                    for (let index = startIndex.y - 1; index >= endIndex.y; index--) {
                        const targetGrid = boardData[index][endIndex.x];
                        if (index === endIndex.y) { // when move to end
                            return true
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            return false
                        }
                    }
                } else if (startIndex.y < endIndex.y) { // to bottom
                    for (let index = startIndex.y + 1; index <= endIndex.y; index++) {
                        const targetGrid = boardData[index][endIndex.x];
                        if (index === endIndex.y) { // when move to end
                            return true
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            return false
                        }
                    }
                }
                return false

            } else if (startIndex.y === endIndex.y) { //level move
                if (startIndex.x > endIndex.x) { //to left
                    for (let index = startIndex.x - 1; index >= endIndex.x; index--) {
                        const targetGrid = boardData[endIndex.y][index];
                        if (index === endIndex.x) {
                            return true
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            return false
                        }
                    }
                } else if (startIndex.x < endIndex.x) { //to right
                    for (let index = startIndex.x + 1; index <= endIndex.x; index++) {
                        const targetGrid = boardData[endIndex.y][index];
                        if (index === endIndex.x) {
                            return true
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            return false
                        }
                    }
                }
                return false

            } else {
                return false
            }
        },
    },
    '2': {
        text: {
            red: '傌',
            black: '馬'
        },
        moveRule: (boardData, startIndex, endIndex, _selfFaction) => {
            const xDifference = endIndex.x - startIndex.x
            const yDifference = endIndex.y - startIndex.y
            const absX = Math.abs(xDifference)
            const absY = Math.abs(yDifference)
            const absDifference = Math.abs(xDifference) + Math.abs(yDifference)
            if (absDifference !== 3) {
                return false
            }
            if (absX === 1) { // vertical move
                if (yDifference > 0) {//to down
                    if (boardData[startIndex.y + 1][startIndex.x] === 'no chessman') {
                        return true
                    }
                    return false// down a grid has obstacle
                } else { //to down
                    if (boardData[startIndex.y - 1][startIndex.x] === 'no chessman') {
                        return true
                    }
                    return false// up a grid has obstacle
                }
            } else if (absY === 1) {//level move
                if (xDifference > 0) {//to right
                    if (boardData[startIndex.y][startIndex.x + 1] === 'no chessman') {
                        return true
                    }
                    return false// right a grid has obstacle
                } else { //to left
                    if (boardData[startIndex.y][startIndex.x - 1] === 'no chessman') {
                        return true
                    }
                    return false// left a grid has obstacle
                }
            } else {
                return false
            }

        }
    },
    '3': {
        text: {
            red: '相',
            black: '象'
        },
        moveRule: (boardData, startIndex, endIndex, selfFaction) => {
            const xDifference = endIndex.x - startIndex.x
            const yDifference = endIndex.y - startIndex.y
            if (Math.abs(xDifference) !== 2 || Math.abs(yDifference) !== 2) { //only move 4 grid
                return false
            }
            if (selfFaction === 'black') { //black
                if (endIndex.y > 4) { //can't cross river
                    return false
                }
            } else { //red
                if (endIndex.y < 5) { //can't cross river
                    return false
                }
            }
            if (xDifference > 0 && yDifference > 0) { //to right down
                if (boardData[startIndex.y + 1][startIndex.x + 1] === 'no chessman') {
                    return true
                }
                return false
            } else if (xDifference > 0 && yDifference < 0) { //to right up
                if (boardData[startIndex.y - 1][startIndex.x + 1] === 'no chessman') {
                    return true
                }
                return false
            } else if (xDifference < 0 && yDifference < 0) { //to left up
                if (boardData[startIndex.y - 1][startIndex.x - 1] === 'no chessman') {
                    return true
                }
                return false
            } else if (xDifference < 0 && yDifference > 0) { //to left down
                if (boardData[startIndex.y + 1][startIndex.x - 1] === 'no chessman') {
                    return true
                }
                return false
            } else {
                return false
            }
        }
    },
    '4': {
        text: {
            red: '仕',
            black: '士'
        },
        moveRule: (_boardData, startIndex, endIndex, selfFaction) => {
            const xDifference = endIndex.x - startIndex.x
            const yDifference = endIndex.y - startIndex.y
            const absX = Math.abs(xDifference)
            const absY = Math.abs(yDifference)
            if (absX > 1 || absY > 1) { //only linear 1 step
                return false
            }
            if (absX + absY !== 2) {
                return false
            }
            if (selfFaction === 'black' && //black
                endIndex.x >= 3 &&
                endIndex.x <= 5 &&
                endIndex.y <= 2 &&
                endIndex.y >= 0) {
                return true
            }
            else if (selfFaction === 'red' && //red
                endIndex.x >= 3 &&
                endIndex.x <= 5 &&
                endIndex.y <= 9 &&
                endIndex.y >= 7) {
                return true
            }
            return false
        }
    },
    '5': {
        text: {
            red: '帥',
            black: '將'
        },
        moveRule: (_boardData, startIndex, endIndex, selfFaction) => {
            const xDifference = startIndex.x - endIndex.x
            const yDifference = startIndex.y - endIndex.y
            const absX = Math.abs(xDifference)
            const absY = Math.abs(yDifference)
            if (absX > 1 || absY > 1) {
                return false
            }
            if (absX + absY !== 1) {
                return false
            }
            if (selfFaction === 'black' && //black
                endIndex.x >= 3 &&
                endIndex.x <= 5 &&
                endIndex.y <= 2 &&
                endIndex.y >= 0) {
                return true
            }
            else if (selfFaction === 'red' && //red
                endIndex.x >= 3 &&
                endIndex.x <= 5 &&
                endIndex.y <= 9 &&
                endIndex.y >= 7) {
                return true
            }
            return false
        }
    },
    '6': {
        text: {
            red: '炮',
            black: '砲'
        },
        moveRule: (boardData, startIndex, endIndex, _selfFaction) => {
            let counter = 0
            let target
            const endChessman = boardData[endIndex.y][endIndex.x]
            if (endChessman === 'no chessman') {
                target = 'no chessman'
            } else {
                target = 'enemy'
            }
            if (startIndex.x === endIndex.x) {  // vertical move
                if (startIndex.y > endIndex.y) { // to up
                    for (let index = startIndex.y - 1; index >= endIndex.y; index--) {
                        const targetGrid = boardData[index][endIndex.x];
                        if (index === endIndex.y) { // when move to end
                            if (target === 'no chessman' && counter === 0) {
                                return true
                            } else if (target === 'enemy' && counter === 1) {
                                return true
                            }
                            return false
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            counter += 1
                        }
                    }
                } else if (startIndex.y < endIndex.y) { // to bottom
                    for (let index = startIndex.y + 1; index <= endIndex.y; index++) {
                        const targetGrid = boardData[index][endIndex.x];
                        if (index === endIndex.y) { // when move to end
                            if (target === 'no chessman' && counter === 0) {
                                return true
                            } else if (target === 'enemy' && counter === 1) {
                                return true
                            }
                            return false
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            counter += 1
                        }
                    }
                }
                return false
            } else if (startIndex.y === endIndex.y) { //level move
                if (startIndex.x > endIndex.x) { //to left
                    for (let index = startIndex.x - 1; index >= endIndex.x; index--) {
                        const targetGrid = boardData[endIndex.y][index];
                        if (index === endIndex.x) {
                            if (target === 'no chessman' && counter === 0) {
                                return true
                            } else if (target === 'enemy' && counter === 1) {
                                return true
                            }
                            return false
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            counter += 1
                        }
                    }
                } else if (startIndex.x < endIndex.x) { //to right
                    for (let index = startIndex.x + 1; index <= endIndex.x; index++) {
                        const targetGrid = boardData[endIndex.y][index];
                        if (index === endIndex.x) {

                            if (target === 'no chessman' && counter === 0) {
                                return true
                            } else if (target === 'enemy' && counter === 1) {
                                return true
                            }
                            return false
                        }
                        if (targetGrid === 'no chessman') {
                            continue
                        } else {
                            counter += 1
                        }
                    }
                }
                return false
            } else {
                return false
            }
        }
    },
    '7': {
        text: {
            red: '兵',
            black: '卒'
        },
        moveRule: (_boardData, startIndex, endIndex, selfFaction) => {
            const xDifference = endIndex.x - startIndex.x
            const yDifference = endIndex.y - startIndex.y
            const absX = Math.abs(xDifference)
            const absY = Math.abs(yDifference)
            if ((absX + absY) > 1) {
                return false
            }
            if (selfFaction === 'black') {
                if (yDifference < 0) { // can't go back
                    return false
                }
                if (endIndex.y < 5 && absX > 0) {
                    return false
                }
            } else if (selfFaction === 'red') {
                if (yDifference > 0) { // can't go back
                    return false
                }
                if (endIndex.y > 4 && absX > 0) {
                    return false
                }
            }
            return true
        }
    },
}
/**
 *  01 => black car
 *  11 => red car
 */
export const startBoard = [
    ['01', '02', '03', '04', '05', '04', '03', '02', '01'],
    ['no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman'],
    ['no chessman', '06', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', '06', 'no chessman'],
    ['07', 'no chessman', '07', 'no chessman', '07', 'no chessman', '07', 'no chessman', '07'],
    ['no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman'],
    ['no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman'],
    ['17', 'no chessman', '17', 'no chessman', '17', 'no chessman', '17', 'no chessman', '17'],
    ['no chessman', '16', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', '16', 'no chessman'],
    ['no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman', 'no chessman'],
    ['11', '12', '13', '14', '15', '14', '13', '12', '11'],
]
export const factionMap: {
    [props: string]: TFaction
} = {
    '0': 'black',
    '1': 'red',
}
export const startX = gridWidth * 2
export const startY = gridWidth * 2
export const tolerance = gridWidth / 2