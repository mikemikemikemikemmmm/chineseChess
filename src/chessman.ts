import Konva from "konva"
import { factionMap, chessmanDataMap, gridWidth } from "./const"
import { TKeyFactionMap, TChessmanData, TFaction, TXYIndex, TMoveRule } from "./types"
export class Chessman {
    private chessmanString: string
    private xyIndex: TXYIndex
    faction!: TFaction
    private categoryData!: TChessmanData
    canva!: Konva.Group
    moveRule: TMoveRule
    constructor(chessmanString: string, xyIndex: TXYIndex) {
        this.chessmanString = chessmanString
        this.xyIndex = xyIndex
        this.setChessmanCategoryData()
        this.moveRule = this.getMoveRule()
        this.setFaction()
        this.createCanva()
    }
    public isKing() {
        return this.chessmanString[1] === '5'
    }
    public setFaction() {
        const factionString = this.chessmanString[0] as TKeyFactionMap
        this.faction = factionMap[factionString]
    }
    public setXYIndex(xyIndex: TXYIndex) {
        this.xyIndex = xyIndex
    }
    public setCanvaPositionByXYIndex() {
        const xyIndex = this.getXYIndex()
        const newPosition = {
            x: xyIndex.x * gridWidth,
            y: xyIndex.y * gridWidth
        }
        this.canva.setPosition(newPosition)
    }
    public getXYIndex() {
        return this.xyIndex
    }
    private setChessmanCategoryData() {
        const categoryString = this.chessmanString[1] as string
        this.categoryData = chessmanDataMap[categoryString]
    }
    public getMoveRule() {
        const moveRule = this.categoryData.moveRule
        if (typeof moveRule === 'function') {
            return moveRule
        }
        return moveRule[this.getFaction()]
    }
    public getFaction() {
        return this.faction
    }
    public destorySelfCanva() {
        this.canva.destroy()
    }
    private createCanva() {
        const group = new Konva.Group({
            draggable: true,
            x: gridWidth * this.xyIndex.x,
            y: gridWidth * this.xyIndex.y
        })
        const circle = new Konva.Circle({
            radius: gridWidth / 3,
            fill: 'white',
            stroke: this.faction,
            strokeWidth: 3
        })
        const text = new Konva.Text({
            text: this.categoryData.text[this.faction],
            fontSize: gridWidth / 2,
            stroke: this.faction,
            strokeWidth: 2,
            x: -gridWidth / 4,
            y: -gridWidth / 4
        })
        group.add(circle)
        group.add(text)
        this.canva =  group
    }
}