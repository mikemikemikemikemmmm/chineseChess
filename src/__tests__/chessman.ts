import { Chessman } from "../chessman"
import { chessmanDataMap } from "../const"
describe("test chessman method", () => {
    const xyIndex = { x: 0, y: 0 }
    const chessmanString = '12' //red horse
    let chessman: Chessman
    beforeAll(() => {
        chessman = new Chessman(chessmanString, xyIndex)
    });
    test('without canva', () => {
        expect(chessman.isKing()).toBe(false)
        expect(chessman.getXYIndex()).toBe(xyIndex)
        expect(chessman.getFaction()).toBe('red')
        expect(chessman.getMoveRule()).toBe(chessmanDataMap['2'].moveRule)
        const newXY = { x: 3, y: 3 }
        chessman.setXYIndex(newXY)
        expect(chessman.getXYIndex()).toBe(newXY)
    })
    test('with canva', () => {
        const dataUrl = chessman.canva.toDataURL()
        expect(dataUrl).toMatchSnapshot()

        chessman.destorySelfCanva()
        const deleteDataUrl = chessman.canva.toDataURL()
        expect(deleteDataUrl).toMatchSnapshot()
    })
})
