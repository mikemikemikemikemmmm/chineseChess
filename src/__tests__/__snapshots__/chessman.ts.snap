// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test chessman method with canva 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAABmJLR0QA/wD/AP+gvaeTAAAF8ElEQVRYhb2Za0hVWRTHf+d0HV+pqWE2lWWFJuZgOVHiB8sMGo3AoOyDCSM004sLTebQwJD4KQimJ9GnSmKix4cZaAbRgR5iT7KnkpUKMzI+KrPsYaPpmg/73nvOuZ57uz7qD5u719prr/3fa7/P1RgjBBKBZGAmMMml7ge6gUfAQw16x9rOSAhNEPhGoFLgXwH5SBoQuC7ws6szn4xYqIBT4J8ASPlKQwLnBZYE2q4WILl84BB2EYiJgfR0mD0bYmOVrr8f2tuhqQkaGmBw0MYlvwKlGnQFStaOWIjAkWGRiI8XKSsTqa8XGRwUv+jpETlzRmTVKhFd945op0DuaMlFC9RZHE6eLHL4sEhfn39SvtDUJFJY6E3yg8Cm0ZC7Z3FUWCjS3T06Yt6oqRGZOtWb6I5AyYVYIqfrIvv2jQ8xM7q6RBYv9l5AG7z5DFskAkeAzQDoOpSXQ17eiEYgYPT1wfbtcOuWRwMs0eC+rb1A/hi2kNGn2Fiz3CAQZEcuVKDVY7hgwecjWFwsEhFh1pW5eekmjt/h3udiY2HtWqMkLg5mzPA9VEFBkJICqamQkWHsh3Z2GRkqmf1NmgQVFWbLn8R1bOqu6Dkwr6KKCggPN8zXr4fSUvtGdV01uHUrFBZCbS3icNjbDgzA0BBcuTLc37ZtkJzslqKA7z0EgRWA6lJ8PJSU2DfgjTlzoKUFrl1TBAFWrkTrch0OK1ZAVhZMn27UuXMH9uwZ7svhgF27zJpvzQTXe9TFxRASEhjBlhaorDTkkhJF1o3sbKirg507DZ3TCbt32/tbtw6iotxSssACN0HjuCkstK8cGWmvLy+Ho0dVfto0yM83yp4+ta/jC6GhsHq1WZOri1oYXwLGwW+H5cuhoMB+kj98aOSXLfNPQsR/eU6OWcrSgXkeMT1dTXpvNDdDUhKcPw9paWoOFRQMt7t0yTrEnZ1w+zY8e2bo3r9XC8UXFi40S/McwCyPOHeufaWuLuUY4MQJleyQkgIPHhjy2bMqmaFp9kGw5zBLB4zJFR1tXykrC3bsUM69ER8Pm1yXkSlT4NAhFW07JCQoX/4QFgbBwW4pWMd8rAQNP2E8yMy0J1hcrCLnRk4OXL5s39m2Nt+buBkGQXTgjUd6+9a+QmMjFBXZz529e6G62qo7eBB6eiyqfk2jOzISamr8kxMx8xAdeOEpbG+3r3TjhjEHIyJgwgQ1nG6HtbWGbUeHIg1qRFJTITGRL0SIffUKJk70T7Cz0/xE6HEAjz2FTU32laKj4Y0r0K9fq9/nz63lbly8CB8+qPyaNXD6tKqzcaPSO53+CVo5PHYAjcAHwEFDA7x8qQ5vM/Ly4Nw5eGEE29NLXbdu7t3dRj4uTv1GRCiigaCuzizd1zV4DdR7GrWbIy0tBrnUVLVKMzKUPDSkzlc3wsKMfG+vYdPWptLHUFVlli66N6Q/PaqTJ4dXunrVyKenw6lTkJtrXz5/vpGvrobr19XZm5AAiYn+I9ncrOwV+gEVLYGZrjeBeoM0NoocOBD4hdPptL43li3zbVtZqWzM/t31t2wx2/4GrtuMBn8DfwBqOKyXR3W+xsTY9zo4GBYtgvp6Izmd1vukG8nJanrU1w8f7tZWOHbMrDlsKRdY7IkiePfm0yanUyQvz6y7YhsMgZMeo8jIz0cwO9ssDwp87eakeRGcAtxD/aphSkryf7iPFb29aniNzXmfBj/4tBfIFfU5QvVo6VKRV6/G/+EuIlJVJRIebo7eTYEvPtopgc2WIUhLE3n0aHzJHTggEhRkJtcqMDXgyAuUWhZNRITI/v0iAwNjI9bS4r0g3OR8XEb9k9wg8M7iLDlZ5MSJkX/hevJE7QwhId7kbo4ocjYkvxL1OcLqOCpKpKhI5Phxkbt3Rd69sxLq6BC5cEGkokIkM1NE07yJDQr8EtCcC4BkkECZwAu/20VwsEh0dCBbS515Kxk3CEQK/CjQNIr97j+B3wWWj6TNgL5R+yC7ENVYFpCCer6a3ww9qLvmXeAS8Jem/poYEf4Hb7ZRLKFiyBgAAAAASUVORK5CYII="`;

exports[`test chessman method with canva 2`] = `"data:,"`;
